### PR TITLE
ENH: Get predictions from Keras models

### PR DESCRIPTION
**Pull request by jaidevd**
This PR allows ModelHandler to accept a Keras model, and draw
predictions from it. See #303 

@sanand0, @bkamapantula and I have been exploring how to make Keras models work with `ModelHandler`. [Here](https://github.com/jaidevd/chart-ann/blob/master/chart_annotation.py#L37) is the subclass we wrote to be able to POST image bytes to the handler and get a predicted label from a pretrained model. As you can see, we barely reused anything in ModelHandler, since almost every method that matters is too tightly coupled with sklearn.

On this branch, we can make the [same app](https://github.com/jaidevd/chart-ann) work with Keras without any subclassing. The changes are still very hacky, though. Need your inputs on how to make `ModelHandler` less sklearn dependent.

### CODE REVIEW

File: gramex/handlers/modelhandler.py [hunk 1]
```diff
@@ -6,16 +6,44 @@
...
+
+        Parameters
+        ----------
+        path : str
+            Path to a Keras model.
+        """
+        model = _model_cache.get(path, False)
+        if not model:
+            _model_cache[path] = gramex.cache.open(path)
+        return _model_cache[path]
```
Comment (pratapvardhan) [inline gramex/handlers/modelhandler.py:30]: @jaidevd -- Since you're using `gramex.cache` you can probably get rid of `_model_cache` another cache check?

File: gramex/handlers/modelhandler.py [hunk 2]
```diff
@@ -6,16 +6,44 @@
...
+
     @classmethod
     def setup(cls, path, **kwargs):
+        if os.path.splitext(path)[-1] == '.h5':
+            from tensorflow.keras.models import load_model
+            gramex.cache.open_callback['h5'] = load_model
```
Comment (pratapvardhan) [inline gramex/handlers/modelhandler.py:36]: @jaidevd -- this would override `.h5` callback across the app. Which means, I can't use `pd.read_hdf`. Instead use 

```python
from tensorflow.keras.models import load_model
cls._model_loader = lambda path: gramex.cache.open(path, load_model)
```
perhaps?

File: gramex/handlers/modelhandler.py [hunk 3]
```diff
@@ -6,16 +6,44 @@
...
+            cls.model_type = "keras"
+            cls._model_loader = cls._load_keras_model
+        else:
+            cls.model_type = "sklearn"
+            cls._model_loader = staticmethod(lambda x: gramex.cache.open(x, gramex.ml.load))
```
Comment (pratapvardhan) [inline gramex/handlers/modelhandler.py:41]: @jaidevd -- should `gramex.ml.load(..., engine='keras')` handle both keras ans sklearn instead here?

Comment (jaidevd) [reply gramex/handlers/modelhandler.py:41]: `gramex.ml.load` is just `joblib.load` - should we add an explicit `gramex.ml.load` that does this (takes the `engine` kwarg and decides what to do)?

**Comment by jaidevd** [comment]
@pratapvardhan Done... with `gramex.cache.open` and `gramex.ml.load`, much smaller change now.

**Review by sanand0 — CHANGES_REQUESTED** [review]

**Comment by jaidevd** [comment]
@sanand0 This is done. External `prepare` methods can be added now. The following config is valid:

```yaml
modelhandler:
    pattern: /$YAMLURL/predictChart/
    handler: ModelHandler
    kwargs:
      path: $YAMLPATH/models/four-class.h5
      prepare: chart_annotation.prepare  # this transforms the binary stream in the request to an np.ndarray
      engine: keras
      xsrf_cookies: false
```

The caveat with Keras models is that we only support image models, and the responsibility of transforming the request body into numpy / tensorflow arrays is on the `prepare` method provided by user. ModelHandler by itself won't do any preprocessing.

We can, however, provide utils that do this in `gramex.ml`, for e.g:

```
from io import BytesIO
from skimage.io import imread

def img_from_bytes(handler):
       handler.request_body['image'] = imread(BytesIO(handler.request.body))
```

which users can use as:
```yaml
modelhandler:
    pattern: /$YAMLURL/predictChart/
    handler: ModelHandler
    kwargs:
      path: $YAMLPATH/models/four-class.h5
      prepare: gramex.ml.img_from_bytes
      engine: keras
      xsrf_cookies: false
```

**Comment by sanand0** [comment]
Closed in favor of a full rewrite of (or replacement for) ModelHandler
